Add quote_identifier function and _to_rawstrings function, fixed the database close connection issue

xiaohuazi123 opened the pull request:
### What does this PR do?

**Add quote_identifier function and _to_rawstrings function, fixed the database close connection issue**

1.Fixed an issue where the database connection was not closed in the tsql_query function.
2.Add the quote_identifier function,used to return the identifier name (column, table, database, and so on) escaped for MSSQL.
3.Add the _to_rawstrings function, to escaped Special characters.


### What issues does this PR fix or reference?
No. 

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [ ] Changelog
- [ ] Tests written/updated

### Commits signed with GPG?
No. 

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

xiaohuazi123 committed:
Add quote_identifier function and _to_rawstrings function, fixed the database close connection issue

(1)Fixed an issue where the database connection was not closed in the tsql_query function.
(2)Add the quote_identifier function,used to return the identifier name (column, table, database, and so on) escaped for MSSQL.
(3)Add the _to_rawstrings function, to escaped Special characters.

xiaohuazi123 committed:
Add quote_identifier function and _to_rawstrings function, fixed the database close connection issue

(1)Fixed an issue where the database connection was not closed in the tsql_query function.
(2)Add the quote_identifier function,used to return the identifier name (column, table, database, and so on) escaped for MSSQL.
(3)Add the _to_rawstrings function, to escaped Special characters.

DmitryKuzmenko commented:
@xiaohuazi123 thank you for contribution. Could you please add some tests. Also since it's not connected to an issue, it would be awesome to see an example of a problem it's fixing.

xiaohuazi123 commented:
> @xiaohuazi123 thank you for contribution. Could you please add some tests. Also since it's not connected to an issue, it would be awesome to see an example of a problem it's fixing.
Hi , @DmitryKuzmenko 
Sorry, I just added an issue to describe the problem to be solved by this PR.
fix  #57195
As for test cases, I don't know how to write them, could you please teach me?

DmitryKuzmenko commented:
@xiaohuazi123 Thank you for update. I think `tests/unit/modules/test_mysql.py` and `tests/unit/modules/test_mysql.py` could be a good example for this.
Also you have lint and pre-commit failures. Could you please install pre-commit hook as described in the [Contribution guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html)? Running pre-commit on your branch will re-format your code and run lint showing the errors to you.
Thanks

xiaohuazi123 commented:
> tests



Hi, @DmitryKuzmenko 
Since I use a Windows system, it seems that pre-commit is not very friendly to Windows, and there are errors. Besides, I also put forward an issue, so I cannot format my code. 
the issue: https://github.com/saltstack/salt/issues/56859

I'm really sorry, for the unit test, I will have a look at your documents first
https://docs.saltstack.com/en/latest/topics/development/tests/index.html#unit-tests
https://docs.saltstack.com/en/latest/topics/development/tests/integration.html
https://docs.saltstack.com/en/latest/topics/development/tests/unit.html

DmitryKuzmenko commented:
Oh I'm very sorry about it. I didn't know that there are issues on Windows with pre-commit.

xiaohuazi123 commented:
> Oh I'm very sorry about it. I didn't know that there are issues on Windows with pre-commit.

Hi @DmitryKuzmenko 
Could you please let me pass lint and pre-commit first, and then I will make up the unit test code in a few days.

DmitryKuzmenko committed:
Blacken salt/modules/mssql.py

DmitryKuzmenko commented:
@xiaohuazi123 I've blackened your code so pre-commit should be good now.
To see the lint failures you can click the "details" link near the lint CI build and then click the "Console Output" and search for ".py:".

xiaohuazi123 commented:
> @xiaohuazi123 I've blackened your code so pre-commit should be good now.
> To see the lint failures you can click the "details" link near the lint CI build and then click the "Console Output" and search for ".py:".

Ok, so I'm going to take care of the lint failures in a couple of days

xiaohuazi123 committed:
Fix too general exception exception error
Fix Undefined variable 'unicode'

xiaohuazi123 committed:
Fix too general exception exception error
Fix Undefined variable 'unicode'

xiaohuazi123 commented:
> @xiaohuazi123 I've blackened your code so pre-commit should be good now.
> To see the lint failures you can click the "details" link near the lint CI build and then click the "Console Output" and search for ".py:".



> @xiaohuazi123 I've blackened your code so pre-commit should be good now.
> To see the lint failures you can click the "details" link near the lint CI build and then click the "Console Output" and search for ".py:".

Hi @DmitryKuzmenko 
Sorry, I forgot to use git rebase, resulting in other commits
Now I have solved the error of lint CI build, please review my code again

DmitryKuzmenko commented:
Pre-commit is failing again. You can view what's wrong in the same way as for lint: https://jenkinsci.saltstack.com/job/pr-pre-commit/job/PR-56764/4/console
There are some strings quoted with single quotes. Black requires double.

xiaohuazi123 committed:
Added the autocommit property to True when getting the database connection.
Commented multiple  conn.autocommit(True)/conn.autocommit(False)  at various points in the code.
Change strings quoted with double quoted.

xiaohuazi123 commented:
> Pre-commit is failing again. You can view what's wrong in the same way as for lint: https://jenkinsci.saltstack.com/job/pr-pre-commit/job/PR-56764/4/console
> There are some strings quoted with single quotes. Black requires double.

Hi @DmitryKuzmenko ,
I have fixed the problem and added the autocommit property to True when getting the database connection. Please help me to review the code again.

DmitryKuzmenko commented:
It looks you've made a mistake managing your branch with git. Now it contains a number of foreign commits. Could you please resolve this?

DmitryKuzmenko commented:
Also please don't comment code lines that you're removing. Just remove them. (I mean commented autocommit on/off)

xiaohuazi123 commented:
> Also please don't comment code lines that you're removing. Just remove them. (I mean commented autocommit on/off)
Hi @DmitryKuzmenko ,
I'm really sorry that I don't know how to solve the problem of foreign commits, which seems to be not solved by git reset.
Or do I open a new branch and resubmit a new pull request?

Also,  I will remove the comment.

DmitryKuzmenko commented:
@xiaohuazi123 you can use `git rebase -i` or create a new branch and `git cherry-pick` your commits to it.

xiaohuazi123 commented:
> create a new branch



Hi @DmitryKuzmenko ,
If I create a new branch and cherry-pick my commits, should I re-launch a pull request?

DmitryKuzmenko commented:
@xiaohuazi123 I think yes.

xiaohuazi123 commented:
> @xiaohuazi123 I think yes.

I feel very sorry, it is my fault, I will now create a new branch and then create a new pull request.

DmitryKuzmenko commented:
But you can do
```
$ git checkout master
$ git pull
$ git checkout Fix-add-close_connection-guote_identifier-in-module-mssql.py
# The next line is to create a backup branch to ensure you'll not lose your current state on rebase. You'll alway able to reset your branch back to the BACKUP-...
$ git branch -c BACKUP-Fix-add-close_connection-guote_identifier-in-module-mssql.py
$ git rebase -i master
```
After this your default editor will be opened with the list of commits to be rebased. You can remove all commits that aren't related to this PR, write and quit from the editor. Gti will start applying the commits you've left in the list. Then you'll need to `git push --force` your branch.

xiaohuazi123 committed:
The autocommit property inside the role_remove function is set to False
fix #57195

xiaohuazi123 commented:
Hi @DmitryKuzmenko ,
I have solved the problem, please kindly review my code again.

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -59,6 +59,10 @@ def __virtual__():
 def _get_connection(**kwargs):
+    """
+    Getting the database connection
+
```
Redundant empty line.

xiaohuazi123 replied:
I'll fix that

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -67,7 +71,71 @@ def _get_connection(**kwargs):
             )
-    return pymssql.connect(**connection_args)
+    conn = pymssql.connect(**connection_args)
+    conn.autocommit(True)
```
Since you're setting `autocommit(True)` here it worth to update all the functions using `_get_connection`.

DmitryKuzmenko replied:
Also `autocommit` could be passed as an argument of `pymssql.connect` so I propose to better put it into args:
```
connection_args['autocommit'] = True
conn = pymssql.connect(**connection_args)
```

xiaohuazi123 replied:
Do you mean it is up to the user to decide whether to open autocommit?
Then set mssql.autocommit: 'true' in the minion configuration file?

DmitryKuzmenko replied:
I don't think so. I'd like to set it after we read all the user input right before `pymssql.connect` is called.

xiaohuazi123 replied:
Ok, I see

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -67,7 +71,71 @@ def _get_connection(**kwargs):
+
+    .. code-block:: bash
+
+        salt '*' mssql.quote_identifier 'foo'bar'
```
Here's a shell syntax error: `'foo'bar'` -- the single quote is not closed. Did you mean either `'foo.bar'` or `"foo'bar"` or something else?

xiaohuazi123 replied:
I'm trying to make the quote_identifier function private, and rename it to _quote_identifier, because it doesn't really make much sense to expose this function to the user

DmitryKuzmenko replied:
This makes sense.

xiaohuazi123 replied:
This should be "foo'bar"

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -96,6 +168,8 @@ def tsql_query(query, **kwargs):
         # Trying to look like the output of cur.fetchall()
         return (("Could not run the query",), (six.text_type(err),))
+    finally:
+        _close_connection(cursor=cur, connect=conn)
```
Could you please also update all the functions here to use `_close_connection` instead of `conn.close` and check it's called in a `finally` block?

xiaohuazi123 replied:
Ok, I'll fix it

DmitryKuzmenko commented:
Hello, @xiaohuazi123 . I left some review comments, could you please follow up?

xiaohuazi123 commented:
> Hello, @xiaohuazi123 . I left some review comments, could you please follow up?


Ok, I'll follow up.

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -67,7 +71,71 @@ def _get_connection(**kwargs):
+    elif six.PY3:
+        if isinstance(text, six.text_type):
+            s = text.encode(encoding="unicode-escape")
+    return s
```
This could be simpler:
```python
s = ""
if isinstance(text, six.text_type):
    s = text.encode(encoding="unicode-escape")
elif isinstance(text, six.string_types):
    s = text.encode(encoding="string-escape")
return s
```
Your don't need to check Py version here because on Py3 text_type is str and string_types contains str only. So you'll not enter the `elif` case on Py3.

xiaohuazi123 replied:
Ok, I see

xiaohuazi123 committed:
fix Redundant empty line.
put autocommit into args.
_close_connection function instead of conn.close
remove the Py3 judgment in _to_rawstrings function

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -428,7 +499,7 @@ def login_remove(login, **kwargs):
         cur.execute("DROP LOGIN [{0}]".format(login))
         conn.autocommit(False)
-        conn.close()
+        _close_connection(cursor=cur, connect=conn)
```
Not in finally block

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -535,7 +606,7 @@ def user_create(
         if conn:
             conn.autocommit(False)
-            conn.close()
+            _close_connection(connect=conn)
```
cursor is not guarded

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -553,12 +624,13 @@ def user_remove(username, **kwargs):
         cur.execute("DROP USER {0}".format(username))
         conn.autocommit(False)
-        conn.close()
+        _close_connection(cursor=cur, connect=conn)
```
not in finally block

DmitryKuzmenko commented:
BTW, I've just checked the mssql API reference, both connection and cursor are context managers so it's better to use `with` statement on them. Could you please update it as well? Let me know if you need an example

xiaohuazi123 commented:
> BTW, I've just checked the mssql API reference, both connection and cursor are context managers so it's better to use `with` statement on them. Could you please update it as well? Let me know if you need an example

Hi @DmitryKuzmenko 
Thank you for your Suggestions. I will fix all of them.
In addition, please give me an example, I am afraid that it will not meet your requirements after modification.
Thank you.

xiaohuazi123 committed:
Change the database connection to the context manager

xiaohuazi123 commented:
> both connection and cursor are context managers so it's better to use `with` statement on them


Hi @DmitryKuzmenko ,
I have changed both connection and cursor to with statement in db_create() function.
Please check to see if it meets your requirements. If it does, then I will change the other functions to the same form.

DmitryKuzmenko commented:
@xiaohuazi123 yes, that looks perfect!

xiaohuazi123 committed:
Change the other functions to change the database connection to the context manager as well

xiaohuazi123 commented:
> @xiaohuazi123 yes, that looks perfect!


Hi @DmitryKuzmenko 

I've changed the database connection for the other functions to a context manager as well.

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -268,24 +340,19 @@ def role_create(role, owner=None, grants=None, **kwargs):
+        with pymssql.connect(_get_db_args(**kwargs)) as conn:
+            with conn.cursor() as cursor:
+                # cur = conn.cursor()
+                # cur.execute(sql)
```
Please remove these commented out code lines.

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -392,23 +457,18 @@ def login_create(
+        with pymssql.connect(_get_db_args(**kwargs)) as conn:
+            with conn.cursor() as cursor:
+                # cur = conn.cursor()
+                # cur.execute(sql)
```
Please remove the commented lines.

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -87,9 +166,11 @@ def tsql_query(query, **kwargs):
+            with conn.cursor() as cursor:
+                cursor.execute(query)
+                # Making sure the result is JSON serializable
         return salt.utils.json.loads(
```
This should be under `with`.

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -87,9 +166,11 @@ def tsql_query(query, **kwargs):
+                cursor.execute(query)
+                # Making sure the result is JSON serializable
         return salt.utils.json.loads(
             _MssqlEncoder().encode({"resultset": cur.fetchall()})
```
Rename `cur` to `cursor`.

DmitryKuzmenko commented:
Also please note that you can check what's wrong with `lint` and `pre-commit` CI builds.

**For lint**
- Click `Details` near `ci/lint`.
- Find the text `Pylint: One warning` and click to `One warning` link.
- You'll see the table with details. Click to `mssql.py:175` the file will be opened with the line highlighted and you'll see the warning details.

**For pre-commit**
- Click `Details` near `ci/pre-commit`
- Click `Console Output`
- Scroll down to `black..........Failed`. You'll see a diff below. You can either apply the changes manually or copy-paste the diff to a file and apply the patch to your code.

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -58,7 +58,26 @@ def __virtual__():
     )
 
 
+def _get_db_args(**kwargs):
```
Oops. Sorry I've missed this point. You don't need this function. You can use with as the following:
```python
with _get_connection(**kwargs) as conn:
```

DmitryKuzmenko commented on salt/modules/mssql.py:
```diff
@@ -67,7 +86,67 @@ def _get_connection(**kwargs):
                 "mssql." + arg, _DEFAULTS.get(arg, None)
             )
-    return pymssql.connect(**connection_args)
+    connection_args['autocommit'] = True
```
Please use double quotes. This is the reason pre-commit is failing.

xiaohuazi123 committed:
Adjust the code format

xiaohuazi123 commented:
> Also please note that you can check what's wrong with `lint` and `pre-commit` CI builds.
> 
> **For lint**
> 
> * Click `Details` near `ci/lint`.
> * Find the text `Pylint: One warning` and click to `One warning` link.
> * You'll see the table with details. Click to `mssql.py:175` the file will be opened with the line highlighted and you'll see the warning details.
> 
> **For pre-commit**
> 
> * Click `Details` near `ci/pre-commit`
> * Click `Console Output`
> * Scroll down to `black..........Failed`. You'll see a diff below. You can either apply the changes manually or copy-paste the diff to a file and apply the patch to your code.



Hi @DmitryKuzmenko 
I have fixed the problem and I will know how to solve it later. 
Thanks for your guidance

xiaohuazi123 committed:
The _close_connection() function is removed

DmitryKuzmenko committed:
Merge branch 'master' into Fix-add-close_connection-quote_identifier-in-module-mssql.py

DmitryKuzmenko commented:
@xiaohuazi123 now code looks good so it's time to write a couple of tests for this. You can use `tests/unit/modules/test_mysql.py` as a good example. Let me know if you'll need help.

xiaohuazi123 commented:
> @xiaohuazi123 now code looks good so it's time to write a couple of tests for this. You can use `tests/unit/modules/test_mysql.py` as a good example. Let me know if you'll need help.


@DmitryKuzmenko 
Ok, I'll make up for unit testing for this in the near future.

xiaohuazi123 committed:
Rewrite the code to just get the database cursor

xiaohuazi123 commented:
> @xiaohuazi123 now code looks good so it's time to write a couple of tests for this. You can use `tests/unit/modules/test_mysql.py` as a good example. Let me know if you'll need help.


Hi , @DmitryKuzmenko
I just changed the code again, changing the database context manager to look like this
```
        with _get_connection(**kwargs).cursor() as cursor:
            cursor.execute(query)
```
But I found that you made another submission on May 28th, so I couldn't push my code to my branch.
Submit id is:898d59ca808a58ce6b4cf496a50c75ee1bd09da3

May I ask if I need to rebase my fork repository before pushing the code?

DmitryKuzmenko commented:
@xiaohuazi123 my commit just merges master branch to this one to actualize your PR. So you cant either rebase your branch to master and `push --force` it. Or reset your branch to the previous version, pull it and apply your change again. Rebase looks easier for me.

Also I'm not sure it is a good idea to do as you described because you lack the connection cleanup. If you would like to use a one-line `with` you have to do like this:
```python
        with _get_connection(**kwargs) as conn, conn.cursor() as cursor:
            cursor.execute(query)
```
In this way both context managers will do the cleanup in the order: cursor first, connection after.

xiaohuazi123 commented:
> @xiaohuazi123 my commit just merges master branch to this one to actualize your PR. So you cant either rebase your branch to master and `push --force` it. Or reset your branch to the previous version, pull it and apply your change again. Rebase looks easier for me.
> 
> Also I'm not sure it is a good idea to do as you described because you lack the connection cleanup. If you would like to use a one-line `with` you have to do like this:
> 
> ```python
>         with _get_connection(**kwargs) as conn, conn.cursor() as cursor:
>             cursor.execute(query)
> ```
> 
> In this way both context managers will do the cleanup in the order: cursor first, connection after.




Before I make up the unit test for this, i need to do  <git rebase xiaohuazi123 fix-add-close_connection-quote_identification-in-module-mssql.py>  branch, right?

DmitryKuzmenko commented:
It's better to do either `git merge master` or `git update; git rebase master` each time you're continuing to work on your branch. Sometimes we could update your PR to the current master when we know we fixed some tests to be sure your branch is free of any related failures. Usually we do not do any change by ourselves without notifying contributors.

So in this particular case it's better to just do rebase on your current version.

xiaohuazi123 committed:
Change the with statement to its original style

xiaohuazi123 commented:
> It's better to do either `git merge master` or `git update; git rebase master` each time you're continuing to work on your branch. Sometimes we could update your PR to the current master when we know we fixed some tests to be sure your branch is free of any related failures. Usually we do not do any change by ourselves without notifying contributors.
> 
> So in this particular case it's better to just do rebase on your current version.
> 
> And thank you for keeping in touch here.




@DmitryKuzmenko 
Ok, thanks for your guidance. I have changed the  `with`  statement to the original style.

dwoz commented:
Closing this due to inactivity. Anyone should feel free to re-open it if they want to see it through to the end in one release cycle.
